### PR TITLE
Fix email validation within candidate_interface/sign_up_form

### PR DESCRIPTION
**Pull request by mnacos**
### Context

We introduced case-insensitive email_address validation in the model but the SignUpForm was doing its own email_address validation and the logic wasn't in sync.

### Changes proposed in this pull request

Refactor SignUpForm so that all email_address validation happens within the model.

### Guidance to review

Run the tests! Read the code. Try to reproduce the bug.

### Link to Trello card

[595 - Can't sign up/in with uppercase version of existing email](https://trello.com/c/Kq1fmTFm)

### Env vars

No new env vars.

### CODE REVIEW

File: spec/system/candidate_interface/candidate_account_spec.rb [hunk 1]
```diff
@@ -40,6 +40,14 @@
 
     when_i_signed_in_more_than_a_week_ago
     then_i_should_be_signed_out
+
+    when_i_click_the_signin_link
+    and_i_submit_my_email_address(uppercase: true)
```
Comment (tijmenb) [inline spec/system/candidate_interface/candidate_account_spec.rb]: These steps should read as English if possible, perhaps this can be:

```suggestion
    and_i_submit_my_email_address_in_uppercase
```

File: app/models/candidate_interface/sign_up_form.rb [hunk 2]
```diff
@@ -1,30 +1,36 @@
...
     validates :email_address, :accept_ts_and_cs, presence: true
-    validates :email_address, length: { maximum: 250 }
-    validates :email_address, email_address: true
+    validate :candidate_email_address_is_valid
 
+    def initialize(params = {})
+      @email_address = params[:email_address]
+      @accept_ts_and_cs = params[:accept_ts_and_cs]
+      @candidate = Candidate.for_email @email_address
+      @candidate ||= Candidate.new email_address: @email_address
```
Comment (tijmenb) [inline app/models/candidate_interface/sign_up_form.rb]: I'm a bit confused by this. `Candidate.for_email` will only return `nil` if `@email_adress` is nil, so this line will return a `Candidate.new` without an email address?

File: app/models/vendor_api_user.rb [hunk 3]
```diff
@@ -1,7 +1,7 @@
 class VendorApiUser < ApplicationRecord
   belongs_to :vendor_api_token
 
-  validates :email_address, presence: true
+  validates :email_address, presence: true, uniqueness: { case_sensitive: false }
```
Comment (tijmenb) [inline app/models/vendor_api_user.rb]: Are we sure these should be unique?

File: app/models/candidate_interface/sign_up_form.rb (hunk 2 above)
Comment (mnacos) [reply app/models/candidate_interface/sign_up_form.rb]: Removing line 13, it is confusing. It was just fixing a couple of rspec validation errors when email_address is nil. I've dealt with this in a different way now, within candidate_email_address_is_valid.

File: app/models/vendor_api_user.rb (hunk 3 above)
Comment (mnacos) [reply app/models/vendor_api_user.rb]: No, will remove for now

File: spec/system/candidate_interface/candidate_account_spec.rb [hunk 4]
```diff
@@ -40,6 +40,14 @@
...
     when_i_signed_in_more_than_a_week_ago
     then_i_should_be_signed_out
+
+    when_i_click_the_signin_link
+    and_i_submit_my_email_address_in_uppercase
+    then_i_receive_an_email_with_a_signin_link
+
+    when_i_visit_the_signup_page
+    and_i_submit_my_email_address_in_uppercase
+    then_i_receive_an_email_with_a_signin_link
```
Comment (mnacos) [inline spec/system/candidate_interface/candidate_account_spec.rb]: @tvararu I added your system spec but changed the L50 to expect a signin link. Is this spec what you'd expect and does it cover the sign-in bug you described?

**Comment by tvararu** [comment]
Tested this locally and it seems to work really well 👍

File: spec/system/candidate_interface/candidate_account_spec.rb (hunk 4 above)
Comment (tvararu) [reply spec/system/candidate_interface/candidate_account_spec.rb]: 👍

File: app/models/reference.rb [hunk 5]
```diff
@@ -3,7 +3,7 @@ class Reference < ApplicationRecord
   validates :email_address, presence: true,
                             email_address: true,
                             length: { maximum: 100 },
-                            uniqueness: { scope: :application_form_id }
+                            uniqueness: { case_sensitive: false, scope: :application_form_id }
```
Comment (tvararu) [inline app/models/reference.rb]: This seems unrelated to the other changes, and doesn't come with a corresponding test, but I think it's fine to sneak it in

**Comment by mnacos** [comment]
> Adding this temporary red review as I want to deploy the current master to prod and then merge this in + do another deploy

Yep that's cool, just make sure you revert this before merging: https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/774/files

**Comment by mnacos** [comment]
> Adding this temporary red review as I want to deploy the current master to prod and then merge this in + do another deploy

Cool, I've rebased this against latest master, so it's ready to be merged when you are ready to deploy.
